### PR TITLE
Added Poker.Evaluate

### PR DESCRIPTION
**Pull request by tonyday567**

### CODE REVIEW

**Comment by tonyday567** [comment]
#9

File: src/Poker/Evaluate.hs [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+-- >>> let cs = [Card Ace Heart,Card Seven Spade,Card Ten Heart,Card Five Spade,Card Six Club, Card Seven Heart,Card Six Spade]
+-- >>> let cs' = [Card Ten Club, Card Five Spade,Card Ten Heart,Card Five Spade,Card Six Club, Card Seven Heart,Card Six Spade]
+-- >>> let css = [cs, cs']
+--
+-- >>> pretty css
+-- [[Ah, 7s, Th, 5s, 6c, 7h, 6s], [Tc, 5s, Th, 5s, 6c, 7h, 6s]]
+
+-- | 5 card standard poker rankings
+--
+-- >>> handRank <$> css
```
Comment (santiweight) [inline src/Poker/Evaluate.hs:60]: Is this the right place for this example? This example seems better placed as a usage at the top of the file or with `handRank` imo

File: src/Poker/Evaluate.hs [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+           not (a == succ b && b == succ c && c == succ d && d == s e),
+           not (a == Ace && [b, c, d, e] == [Five, Four, Three, Two])
+       ]
+    ++ [FullHouse a b | a <- ranks, b <- ranks, a /= b]
+    ++ [FourOfAKind a b | a <- ranks, b <- ranks, a /= b]
+    ++ [StraightFlush f | f <- ranksGE Five]
+  where
+    s Ace = Two
+    s other = succ other
+    ranks = [Two .. Ace]
```
Comment (santiweight) [inline src/Poker/Evaluate.hs:129]: `allRanks` - can also inline after that.

File: src/Poker/Evaluate.hs [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+    ranksLT Two = []
```
Comment (santiweight) [inline src/Poker/Evaluate.hs:130]: `ranksLT = flip takeWhile allRanks . (/=)`

codegolf lol

File: src/Poker/Evaluate.hs [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+    ranksLT rank = [Two .. pred rank]
+    ranksGE rank = reverse [Ace, King .. rank]
```
Comment (santiweight) [inline src/Poker/Evaluate.hs:132]: I don't use `..` lists very often. Does `ranksGE Ace` return `[]`?

File: src/Poker/Evaluate.hs [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+-- | enumeration of all possible HandRanks, in ascending order.
+--
+-- >>> length allHandRanks
+-- 7462
+--
+allHandRanks :: [HandRank]
+allHandRanks =
+  [ HighCard a b c d e
+    | a <- ranks,
+      b <- ranksLT a,
```
Comment (santiweight) [inline src/Poker/Evaluate.hs:83]: I wonder whether we should expose `ranksLT` in `Cards`? There's a few usages of that logic in `Cards` currently, and this seems like a useful function...

File: src/Poker/Evaluate.hs [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+       ]
+    ++ [Straight f | f <- ranksGE Five]
+    ++ [ Flush a b c d e
+         | a <- ranks,
+           b <- ranksLT a,
+           c <- ranksLT b,
+           d <- ranksLT c,
+           e <- ranksLT d,
+           not (a == succ b && b == succ c && c == succ d && d == s e),
+           not (a == Ace && [b, c, d, e] == [Five, Four, Three, Two])
```
Comment (santiweight) [inline src/Poker/Evaluate.hs:121]: `isNothing $ straight [a,b,c,d,e]`

File: poker-base.cabal [hunk 7]
```diff
@@ -27,6 +27,7 @@ source-repository head
 library
   exposed-modules:
     Poker
+    Poker.Evaluate
```
Comment (santiweight) [inline poker-base.cabal:30]: How about HandRank? I think Evaluate is quite an overloaded word (like monte-carlo equity for example)

**Review by santiweight — COMMENTED** [review]
Just some small thoughts. I didn't really check the logic, which I would think would be a bad idea before we have some tests to check that the logic is in fact correct. I think a good way to get good tests might be to use a source of truth - some online API that wouldn't mind us harrassing them, or a C++ library.

What do you think about the feasability of such tests?

**Comment by santiweight** [comment]
Issue regarding tests for `Evaluate` (modulo naming) 

https://github.com/santiweight/poker-base/issues/30

**Comment by tonyday567** [comment]
I'd like to clean it up first and add the tests.  This is not for the initial release, right?

**Pull request closed without merging**